src/Compilers/Core/Portable/Text/SourceText.cs

```diff
@@ -660,26 +658,60 @@ ImmutableArray<byte> computeContentHash()
             }
         }
 
-        internal static ImmutableArray<byte> CalculateChecksum(byte[] buffer, int offset, int count, SourceHashAlgorithm algorithmId)
+#if NETCOREAPP
+        internal static ImmutableArray<byte> CalculateChecksum(ReadOnlySpan<byte> buffer, SourceHashAlgorithm algorithmId)
         {
-            using (var algorithm = CryptographicHashProvider.TryGetAlgorithm(algorithmId))
+            using var algorithm = CryptographicHashProvider.GetAlgorithm(algorithmId);
+            int size = algorithmId switch
+            {
+                SourceHashAlgorithm.Sha1 => SHA1.HashSizeInBytes,
+                SourceHashAlgorithm.Sha256 => SHA256.HashSizeInBytes,
+                _ => throw ExceptionUtilities.UnexpectedValue(algorithmId)
+            };
+
+            Span<byte> hashBuffer = stackalloc byte[size];
+            if (!algorithm.TryComputeHash(buffer, hashBuffer, out int bytesWritten))
             {
-                RoslynDebug.Assert(algorithm != null);
-                return ImmutableArray.Create(algorithm.ComputeHash(buffer, offset, count));
+                throw new InvalidOperationException();
             }
+
+            return ImmutableArray.Create(hashBuffer.Slice(0, bytesWritten));
+        }
+#endif
+
+        internal static ImmutableArray<byte> CalculateChecksum(byte[] buffer, int offset, int count, SourceHashAlgorithm algorithmId)
+        {
+#if NETCOREAPP
+            return CalculateChecksum(new ReadOnlySpan<byte>(buffer, offset, count), algorithmId);
+#else
+            using var algorithm = CryptographicHashProvider.GetAlgorithm(algorithmId);
```

Comment:
Honestly, nothing in this path is striking me as less efficient that the NETCOREAPP version. What is the perf difference that we'd want to have the other one?